File: behavioral/visitor/code_analyzer/python/code_analyzer/ast_nodes.py
from abc import ABC, abstractmethod
from .visitor import Visitor
from typing import List, Optional

# --- Element Interface ---
class Node(ABC):
    """Element interface that declares an accept method."""
    @abstractmethod
    def accept(self, visitor: Visitor):
        pass

# --- Concrete Elements (AST Nodes) ---
class ExpressionNode(Node):
    """Represents a simple code expression (e.g., literal, variable, binary op)."""
    def __init__(self, representation: str):
        self.representation = representation # Simplified representation for the demo

    def accept(self, visitor: Visitor):
        if hasattr(visitor, 'visit_expression'):
            visitor.visit_expression(self)


class StatementNode(Node):
    """Base class for statement nodes."""
    pass


class FunctionDefinitionNode(Node):
    """Represents a function definition."""
    def __init__(self, name: str, parameters: List[str], body: List[StatementNode]):
        self.name = name
        self.parameters = parameters
        self.body = body

    def accept(self, visitor: Visitor):
        visitor.visit_function_definition(self)


class VariableDeclarationNode(StatementNode):
    """Represents a variable declaration (e.g., var x: int = 5)."""
    def __init__(self, name: str, type_hint: str, initializer: Optional[ExpressionNode] = None):
        self.name = name
        self.type_hint = type_hint
        self.initializer = initializer

    def accept(self, visitor: Visitor):
        visitor.visit_variable_declaration(self)


class AssignmentStatementNode(StatementNode):
    """Represents an assignment (e.g., x = 10)."""
    def __init__(self, target_variable: str, value: ExpressionNode):
        self.target_variable = target_variable
        self.value = value

    def accept(self, visitor: Visitor):
        visitor.visit_assignment_statement(self)


class IfStatementNode(StatementNode):
    """Represents an if(-else) statement."""
    def __init__(self, condition: ExpressionNode, then_branch: List[StatementNode], else_branch: Optional[List[StatementNode]] = None):
        self.condition = condition
        self.then_branch = then_branch
        self.else_branch = else_branch

    def accept(self, visitor: Visitor):
        visitor.visit_if_statement(self)


class ExpressionStatementNode(StatementNode):
    """Represents a statement that consists of just an expression (e.g., a function call)."""
    def __init__(self, expression: ExpressionNode):
        self.expression = expression

    def accept(self, visitor: Visitor):
        visitor.visit_expression_statement(self)

File: behavioral/visitor/code_analyzer/typescript/code_analyzer/ast_nodes.ts
import { Visitor } from "./visitor";

// --- Element Interface ---
export interface Node {
  // Element interface that declares an accept method.
  accept(visitor: Visitor): void;
}

// --- Concrete Elements (AST Nodes) ---

// Represents a simple code expression (e.g., literal, variable, binary op).
export class ExpressionNode implements Node {
  // Simplified representation for the demo
  constructor(public representation: string) {}

  accept(visitor: Visitor): void {
    // If visitors need to specifically handle expressions, they might implement visitExpression
    // visitor.visitExpression?.(this); // Use optional chaining if method is optional
    // Keep simple: traversal usually happens from statements containing expressions
  }
}

// Base class/interface for statement nodes. Using an abstract class for potential common methods.
export abstract class StatementNode implements Node {
  abstract accept(visitor: Visitor): void;
}

// Represents a function definition.
export class FunctionDefinitionNode implements Node {
  constructor(
    public name: string,
    public parameters: string[],
    public body: StatementNode[]
  ) {}

  accept(visitor: Visitor): void {
    visitor.visitFunctionDefinition(this);
    // Traversal is now the responsibility of the visitor's visit method.
  }
}

// Represents a variable declaration (e.g., var x: int = 5).
export class VariableDeclarationNode extends StatementNode {
  constructor(
    public name: string,
    public typeHint: string,
    public initializer?: ExpressionNode
  ) {
    super();
  }

  accept(visitor: Visitor): void {
    visitor.visitVariableDeclaration(this);
    this.initializer?.accept(visitor); // Visit initializer if it exists
  }
}

// Represents an assignment (e.g., x = 10).
export class AssignmentStatementNode extends StatementNode {
  constructor(public targetVariable: string, public value: ExpressionNode) {
    super();
  }

  accept(visitor: Visitor): void {
    visitor.visitAssignmentStatement(this);
    this.value.accept(visitor);
  }
}

// Represents an if(-else) statement.
export class IfStatementNode extends StatementNode {
  constructor(
    public condition: ExpressionNode,
    public thenBranch: StatementNode[],
    public elseBranch?: StatementNode[] // Optional else branch
  ) {
    super();
  }

  accept(visitor: Visitor): void {
    visitor.visitIfStatement(this);
    // Traversal is now the responsibility of the visitor's visit method.
  }
}

// Represents a statement that consists of just an expression (e.g., a function call).
export class ExpressionStatementNode extends StatementNode {
  constructor(public expression: ExpressionNode) {
    super();
  }

  accept(visitor: Visitor): void {
    visitor.visitExpressionStatement(this);
    this.expression.accept(visitor);
  }
}
